[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.12669922.svg)](https://doi.org/10.5281/zenodo.12669922)

This archive includes all processed data that have been generated and used in the paper 'Variation in the fitness impact of translationally optimal codons among animals', as well as the scripts used to analyze the data and to generate the figures.

Developed by Florian Bénitière, Tristan Lefébure and Laurent Duret. Laboratoire de Biométrie et Biologie Évolutive, Université Lyon 1, UMR CNRS 5558, Villeurbanne, France. Laboratoire d'Ecologie des Hydrosystèmes Naturels et Anthropisés, Université Lyon 1, UMR CNRS 5023, Villeurbanne, France.

### Directories content

-   The folder '**figure**' contains the necessary materials to produce the figures of the manuscript:

    -   Intermediate pannels generated by R scripts to produce figures are stored in the directory '**pannels**'.

    -   Stable images used in pannels are stored in '**images_library**'.

    -   Directories labeled with the extension '**\_generator**' contain R scripts used to produce figures.

    -   The remaining directories are where the figures are stored.

-   The '**data**' directory contains processed data in tab-separated text format, which is used for generating figures and conducting analyses. The script to produce this table are located in the '**pipelines**' folder.

    -   The '**per_species**' directory contains compressed tab-separated text format files for each of the 257 studied species. Each species possesses a directory '*`Scientific name`*\_NCBI.taxid*`taxID`*', within can be found a subdirectory with the *`assembly accession`*.

        -   '`codon_usage_gene_fpkm.txt.gz`' encompasses the occurrences of codons within exonic regions and triplet sequences within intronic regions, assessed for each coding sequence. Gene expression metrics, including mean_fpkm, median_fpkm, or weighted_fpkm, are derived from GTDrift (Bénitière et al. 2024) and compiled for analysis.

        -   '`decoding_table.tab.gz`' contains information regarding the gene copy count per tRNA. Additionally, it delineates whether a codon is translated via Watson-Crick base pairing with the most prevalent tRNA for a given amino acid or through wobble pairing, which occurs when the corresponding tRNA is absent.

        -   '`trna_from_gff.txt.gz`' consist of annotated copies obtained from the GFF files downloaded from NCBI.

        -   '`trna_from_trnascanse.txt.gz`' comprises annotated copies generated by the tRNAscan-SE tool, executed through the command line:

            ```         
            tRNAscan-SE -E -o trna_from_trnascanse.txt -S off genome.fna
            ```

    -   '`data1_supp.tab`' Contains multiple codon usage metrics for various species, utilized in Figures 1, 2, 4, 5, 6, 7, and Supplementary Figures 1, 3, 4, 5, 7, 8, 9.

    -   '`data2_supp.tab`' Includes per-gene GCi and GC3 data for selected species, used in Figure 1.

    -   '`data3_supp.tab`' Provides per-amino acid data on tRNA gene copy number and amino acid usage for selected species, used in Figure 2.

    -   '`data4_supp.tab`' Contains per-species data on tRNA gene copy number, used in Figure 3.

    -   '`data5_supp.tab`' Details the frequency of putative-optimal codons by 2% bins of gene expression level for selected species, used in Figure 4 and Supplementary Figures 6, 10.

    -   '`data6_supp.tab`' Contains data on sites constraint per gene for selected species, used in Supplementary text Figure 3.

    -   '`data7_supp.tab`' Provides per-species data on tRNA gene copy number, used in Supplementary Figure 2.

    -   '`data8_supp.tab`' Includes per-species data on correlation with the tRNA gene copy number of *Hydra vulgaris*, used in Supplementary Figure 3.

    -   '`data9_supp.tab`' Contains data on tRNA gene copy number per amino acid for *Caenorhabditis elegans* and *Hydra vulgaris*, used in Supplementary Figure 3.

    -   '`data10_supp.tab`' Provides data from Behrens et al. 2021 on tRNA abundance, gene copy number, and amino acid usage, used in Supplementary Figure 4.

    -   '`data11_supp.tab`' Contains data on tRNA gene copy number for dipters and lepidopters, similar to '`data4_supp.tab`', used in Supplementary Figure 8.

    -   '`data12_supp.tab`' Details Valine synonymous codon usage with gene expression level for selected species, used in Supplementary Figure 9.

    -   '`data13_supp.tab`' Includes data on single nucleotide polymorphisms of putative-optimal codons and non-putative-optimal codons alleles, used in Supplementary text Figure 1.

    -   '`data14_supp.tab`' Contains data on substitutions of putative-optimal codons and non-putative-optimal codons alleles, used in Supplementary text Figure 1.

    -   '`data15_supp.tab`' Provides per-species data on GCi and GC3 along the genes, used in Supplementary text Figure 2.

    -   '`data16_supp.tab`' Contains data on codon usage between highly expressed genes and lowly expressed genes for dipters and lepidopters, used in the text to measure values, refer to '`values_in_text_paper_generator.R`').

-   The '**pipelines**' folder contains the bionformatics pipelines for three different purposes: to download data from repository ('**building_from_source**'); to analyze various characteristics as part of our study ('**workflow**'); to generate data table located in the '**data**' directory ('**data_generator**').

The script '`values_in_text_paper_generator.R`' retrieves the values found in the main text of the paper from the data.

Some packages that you may encounter in these R scripts should be install: - stringr - kableExtra - cowplot - ggplot2 - imager - RColorBrewer ...

#### References

Florian Bénitière, Laurent Duret, Anamaria Necsulea, GTDrift: a resource for exploring the interplay between genetic drift, genomic and transcriptomic characteristics in eukaryotes, *NAR Genomics and Bioinformatics*, Volume 6, Issue 2, June 2024, lqae064, <https://doi.org/10.1093/nargab/lqae064>
