[![DOI](https://zenodo.org/)](https://zenodo.org/)

This archive includes all processed data that have been generated and used in the paper 'Non-adaptive mechanisms are the primary drivers of codon usage in metazoans', as well as the scripts used to analyze the data and to generate the figures.

Developed by Florian Bénitière, Tristan Lefébure and Laurent Duret. Laboratoire de Biométrie et Biologie Évolutive, Université Lyon 1, UMR CNRS 5558, Villeurbanne, France. Laboratoire d'Ecologie des Hydrosystèmes Naturels et Anthropisés, Université Lyon 1, UMR CNRS 5023, Villeurbanne, France.

### Directories content

-   The folder 'figure' contains the necessary materials to produce the figures of the manuscript:

    -   Intermediate pannels generated by R scripts to produce figures are stored in the directory 'pannels'.

    -   Stable images used in pannels are stored in 'images_library'.

    -   Directories labeled with the extension '\_generator' contain R scripts used to produce figures.

    -   The remaining directories are where the figures are stored.

-   The 'data' directory contains processed data in tab-separated text format, which is used for generating figures and conducting analyses.

    -   The 'per_species' directory contains compressed tab-separated text format files for each of the 257 studied species. Each species possesses a directory '*`Scientific name`*\_NCBI.taxid*`taxID`*', within can be found a subdirectory with the *`assembly accession`*.

        -   'codon_usage_gene_fpkm' which contain coding sequence the number occurences of each codons within exons, and each triplets within introns. The gene expression value such as mean_fpkm, median_fpkm or weighted_fpkm are collected from GTDrift.

        -   'decoding_table.tab.gz' contains the number of gene copy *per* tRNA. Also it indicates if a codon is decoded through Watson-Crick of the most abundant tRNA of an amino acid, or through wobble pairing.

        -   'tRNAscan.tab.gz' annotated copies from the GFF files download from NCBI.

        -   'tRNAscan_SE.tab.gz' annotated copies by tRNAscan-SE
        
        

    -   'Data1_supp.tab' encompasses per species data on life history traits, number of identified genes among the eukaryota BUSCO gene set, and the number of genes that have been associated to one annotated protein-coding gene only (Fig. 1,2,3,5,6,7,8).

-   The 'pipelines' folder contains the bionformatics pipelines for different purposes.

    -   To calculate the dN/dS ratio ('dNdS_pipeline').

### 

Some packages that you may encounter in these R scripts should be install: - stringr - kableExtra - cowplot - ggplot2 - imager - RColorBrewer ...

### 

The script 'values_in_text_paper_generator.R' retrieves the values found in the main text of the paper from the data.
